docs/_design/map-controller.md

```diff
@@ -0,0 +1,399 @@
+SnapshotSchedule controller:
+
+```yaml
+apiVersion: metacontroller.k8s.io/v1alpha1
+kind: MapController
+metadata:
+  name: snapshotschedule-controller
+spec:
+  parentResource:
+    apiVersion: snapshot.k8s.io/v1
+    resource: snapshotschedules
+  inputResources:
+  - apiVersion: v1
+    resource: persistentvolumeclaims
+  outputResources:
+  - apiVersion: volumesnapshot.external-storage.k8s.io/v1
+    resource: volumesnapshots
+  resyncPeriodSeconds: 5
+  hooks:
+    map:
+      webhook:
+        url: http://snapshotschedule-controller.metacontroller/map
+    tombstone:
+      webhook:
+        url: http://snapshotschedule-controller.metacontroller/tombstone
+```
+
+### Parent Resource
+
+The parent resource is the SnapshotSchedule itself, and anything this controller
+creates will by owned by this parent.
+The schedule thus acts like a bucket containing snapshots: if you delete the
+schedule, the snapshots inside it will go away too, unless you use
+`--cascade=false` to orphan them.
+Notably, this ties the lifecycles of snapshots to the reason they exist
+(the backup policy that the user defined), rather than tying them to the entity
+that they are about (the PVC).
+
+### Input Resources
+
+The input resources (in this case just PVC) are the inputs to the conceptual
+"map" function.
+We allow multiple input resources because users might want to write a controller
+that performs the same action for several different input types.
+We shouldn't force them to create multiple MapControllers with largely identical
+behavior.
+
+The duck-typed `spec.selector` field (assumed to be `metav1.LabelSelector`) in
+the parent object is used to filter which input objects to process.
+If the selector is empty, we will process all objects of the input types in the
+same namespace as the parent.
+
+If there are multiple input resources, they are processed independently, with no
+attempt to correlate them.
+That is, the [map hook][] will still be called with only a single input object
+each time, although the kind of that object might be different from one call to
+the next.
+
+[map hook]: #map-hook
+
+### Output Resources
+
+The output resources (in this case just VolumeSnapshot) are the types of objects
+that the user intends to create and hold in the conceptual "bucket" that the
+parent object represents.
+We allow multiple output resources because users might think of their controller
+as spitting out a few different things.
+We shouldn't force them to create a CompositeController too just so they can
+emit multiple outputs, especially if those outputs are not conceptually part of
+one larger whole.
+
+For a given input object, the user can generate any number of output objects.
+We will [tag](#map-key) those output objects in some way to associate them with
+the object that we sent as input.
+The tag makes it possible to group those objects and send them along with future
+[map hook requests](#map-hook-request).
```

Comment:
I'm not sure if I understand the MxN input/output case. IIUC, we tag each output project on which input object they'll be generated from?

Reply:
On the input side, the fact that there can be multiple input resources is only used for selecting the pool of input objects. In pseudocode, something like this:

```go
inputObjects := []Object{}
for _, inputResource := range inputResources {
  inputObjects = append(inputObjects, getMatchingObjects(inputResource, parentSelector)...)
}
```

Then we go through each of the input objects and call the user's hook to generate the output objects for that one input. We tag each output with the input that it came from, which also allows us to select the observed outputs belonging to this input. Continuing in pseudocode:

```go
outputObjects := []Object{}
for _, inputObject := range inputObjects {
  // Compute some opaque string identifying this input object.
  mapKey := makeMapKey(inputObject)

  // Gather observed objects of the output resources that are tagged with this key.
  observedOutputs := []Object{}
  for _, outputResource := range outputResources {
    // Gather all outputs owned by this parent.
    allOutputs := getOwnedObjects(outputResource, parent)
    // Filter to only those tagged for this input.
    observedOutputs = append(observedOutputs, filterByMapKey(allOutputs, mapKey)...)
  }

  // Call user's map hook, passing observed state.
  mapResult := mapHook(parent, inputObject, observedOutputs)
  for _, obj := range mapResult.Outputs {
    // Tag outputs to identify which input they came from.
    setMapKey(obj, mapKey)
    outputObjects = append(outputObjects, obj)
  }
}
reconcileObjects(outputObjects)
```

Does that help? If so, I can add this to the doc somewhere.

Reply:
Yes, this helps.